src/client/opamArg.ml

```diff
@@ -782,62 +774,414 @@ let opamlist_columns =
+        let d = update_doc_w_cli d ~cli validity in
+        `I (cmds, d)
+      ) commands
+
+  let mk_subcommands_aux ~cli my_enum commands =
+    let commands =
+      List.map (fun (v,n,c,a,d) -> contented_validity v c, n, a, d) commands
+    in
+    let command =
+      let doc = Arg.info ~docv:"COMMAND" [] in
+      let scommand = List.rev_map (fun (v,f,_,_) -> f,v.content) commands in
+      let check = function
+        | None -> `Ok None
+        | Some elem ->
+          match OpamStd.List.find_opt (fun (validity, _, _, _) ->
+              validity.content = elem) commands with
+          | Some (validity, sbcmd, _,_) ->
+            check_cli_validity cli validity (Some (elem_of_vr elem)) [sbcmd]
+          | None -> `Ok (Some (elem_of_vr elem))
+      in
+
+      term_cli_check ~check Arg.(pos 0 (some & my_enum scommand) None & doc)
+    in
+    let params =
+      let doc = Arg.info ~doc:"Optional parameters." [] in
+      Arg.(value & pos_right 0 string [] & doc)
+    in
+    command, params
+
+  let mk_subcommands ~cli commands =
+    mk_subcommands_aux ~cli Arg.enum commands
+
+  let mk_subcommands_with_default ~cli commands =
+    let enum_with_default_valrem sl =
+      let parse, print = Arg.enum sl in
+      let parse s =
+        match parse s with
+        | `Ok x -> `Ok (x)
+        | _ -> `Ok (`Valid (`default s)) in
+      parse, print
+    in
+    mk_subcommands_aux ~cli enum_with_default_valrem commands
+
+  let bad_subcommand ~cli subcommands (command, usersubcommand, userparams) =
+    match usersubcommand with
+    | None ->
+      `Error (false,
+              Printf.sprintf "Missing subcommand. Valid subcommands are %s."
+                (OpamStd.Format.pretty_list
+                   (OpamStd.List.filter_map (fun (validity,sb,_,_,_) ->
+                        match validity with
+                        | {valid = c; removed = None; _} when c < cli -> None
+                        | {removed = Some (c,_); _}  when c >= cli -> None
+                        | _ -> Some sb)
+                       subcommands)))
+    | Some (`default cmd) ->
+      `Error (true, Printf.sprintf "Invalid %s subcommand %S" command cmd)
+    | Some usersubcommand ->
+      let exe = Filename.basename Sys.executable_name in
+      match
+        List.find_all (fun (_,_,cmd,_,_) ->
+            cmd = usersubcommand) subcommands
+      with
+      | [ _, name, _, args, _doc] ->
+        let usage =
+          Printf.sprintf "%s %s [OPTION]... %s %s"
+            exe command name (String.concat " " args) in
+        if List.length userparams < List.length args then
+          `Error (false, Printf.sprintf "%s: Missing argument.\nUsage: %s\n"
+                    exe usage)
+        else
+          `Error (false, Printf.sprintf "%s: Too many arguments.\nUsage: %s\n"
+                    exe usage)
+      | _ ->
+        `Error (true, Printf.sprintf "Invalid %s subcommand" command)
+
+end
+
+include Mk
```

Comment:
Does it need to be a module, therefore?

Reply:
It is not programmatically needed. I wanted to enclose all internal cli mechanism into the module, in order to forbid use of validity types and functions, and access them only via exposed ones `mk_*` functions.